### PR TITLE
Add rest api for create/delete repo hook

### PR DESCRIPTION
**Pull request by j0nathan33**

### CODE REVIEW

File: src/main/java/com/cdancy/bitbucket/rest/domain/repository/Hook.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.List;
+
+@AutoValue
+public abstract class Hook implements ErrorsHolder {
+
+    @Nullable
+    public abstract HookDetails details();
+
+    @Nullable
```
Comment (cdancy) [inline src/main/java/com/cdancy/bitbucket/rest/domain/repository/Hook.java]: Lets not make boolean values Nullable. Instead use lower-case boolean here and within the `create` method, you can do something inline like: `Boolean.valueOf(passValue)`

File: src/main/java/com/cdancy/bitbucket/rest/domain/repository/Hook.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    public abstract Boolean enabled();
+
+    @Nullable
+    public abstract Boolean configured();
```
Comment (cdancy) [inline src/main/java/com/cdancy/bitbucket/rest/domain/repository/Hook.java]: Same here.

File: src/main/java/com/cdancy/bitbucket/rest/features/RepositoryApi.java [hunk 3]
```diff
@@ -147,4 +149,46 @@ PermissionsPage listPermissionsByGroup(@PathParam("project") String project,
...
                                          @Nullable @QueryParam("start") Integer start,
                                          @Nullable @QueryParam("limit") Integer limit);
+
+    @Named("repository:list-Hook")
+    @Documentation({"https://developer.atlassian.com/static/rest/bitbucket-server/5.0.1/bitbucket-rest.html#idm45993794419936"})
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Path("/{project}/repos/{repo}/settings/hooks")
+    @Fallback(BitbucketFallbacks.HookPageOnError.class)
+    @GET
+    HookPage listHook(@PathParam("project") String project,
```
Comment (cdancy) [inline src/main/java/com/cdancy/bitbucket/rest/features/RepositoryApi.java]: Change to `listHooks`?

File: src/main/java/com/cdancy/bitbucket/rest/domain/repository/Hook.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    public abstract HookDetails details();
+
+    public abstract boolean enabled();
+
+    public abstract boolean configured();
+
+    @SerializedNames({ "details", "enabled", "configured", "errors" })
+    public static Hook create(HookDetails details, String enabled, String configured, @Nullable List<Error> errors) {
+        return new AutoValue_Hook(Utils.nullToEmpty(errors), details, Boolean.valueOf(enabled), Boolean.valueOf(configured));
+    }
```
Comment (cdancy) [inline src/main/java/com/cdancy/bitbucket/rest/domain/repository/Hook.java]: Think you misunderstood ;) Lets do something like this:

```
@SerializedNames({ "details", "enabled", "configured", "errors" })
 public static Hook create(HookDetails details, Boolean enabled, Boolean configured, @Nullable List<Error> errors) {
         return new AutoValue_Hook(Utils.nullToEmpty(errors), details, Boolean.valueOf(enabled), Boolean.valueOf(configured));
 }
```

This way, if those values are null, it will still resolve to false for the actual object creation.

Comment (j0nathan33) [inline src/main/java/com/cdancy/bitbucket/rest/domain/repository/Hook.java]: I try with boolean and I have Exception : "java.lang.NullPointerException". I try with main and I have same error : 

> public static void main(String[] args) {
>         Boolean test = null;
>         System.out.println(Boolean.valueOf(test));
>     }

Comment (cdancy) [inline src/main/java/com/cdancy/bitbucket/rest/domain/repository/Hook.java]: Just did the same and was able to do so successfully. Printed out `false` in this case.

```
    
    public static void main(String [] args) {
        String fish = null;
        boolean success = Boolean.valueOf(fish);
        System.out.println(success);
    }
```

Comment (cdancy) [inline src/main/java/com/cdancy/bitbucket/rest/domain/repository/Hook.java]: Ok found the issue. Code the class like so:

```
@AutoValue
public abstract class Hook implements ErrorsHolder {

    @Nullable
    public abstract HookDetails details();

    public abstract boolean enabled();

    public abstract boolean configured();

    @SerializedNames({ "details", "enabled", "configured", "errors" })
    public static Hook create(HookDetails details, boolean enabled, boolean configured, @Nullable List<Error> errors) {
        return new AutoValue_Hook(Utils.nullToEmpty(errors), details, enabled, configured);
    }
}
```
You'll have to then fix the compilation issue in the fallbacks class which you can just pass in false.

Comment (cdancy) [inline src/main/java/com/cdancy/bitbucket/rest/domain/repository/Hook.java]: I forgot that booleans will be magically resolved correctly and we don't need to just through any hoops and/or set Nullable. Just have to believe it will work :)

File: src/main/java/com/cdancy/bitbucket/rest/features/RepositoryApi.java [hunk 5]
```diff
@@ -167,4 +169,46 @@ PermissionsPage listPermissionsByGroup(@PathParam("project") String project,
...
+                    @PathParam("repo") String repo,
+                    @PathParam("hookKey") String hookKey);
+
+    @Named("repository:create-Hook")
+    @Documentation({"https://developer.atlassian.com/static/rest/bitbucket-server/5.0.1/bitbucket-rest.html#idm45993794409760"})
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Path("/{project}/repos/{repo}/settings/hooks/{hookKey}/enabled")
+    @Fallback(BitbucketFallbacks.HookOnError.class)
+    @PUT
+    Hook createHook(@PathParam("project") String project,
```
Comment (cdancy) [inline src/main/java/com/cdancy/bitbucket/rest/features/RepositoryApi.java]: Lets rename to `enableHook`

File: src/main/java/com/cdancy/bitbucket/rest/features/RepositoryApi.java [hunk 6]
```diff
@@ -167,4 +169,46 @@ PermissionsPage listPermissionsByGroup(@PathParam("project") String project,
...
+                    @PathParam("repo") String repo,
+                    @PathParam("hookKey") String hookKey);
+
+    @Named("repository:delete-Hook")
+    @Documentation({"https://developer.atlassian.com/static/rest/bitbucket-server/5.0.1/bitbucket-rest.html#idm45993794409760"})
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Path("/{project}/repos/{repo}/settings/hooks/{hookKey}/enabled")
+    @Fallback(BitbucketFallbacks.HookOnError.class)
+    @DELETE
+    Hook deleteHook(@PathParam("project") String project,
```
Comment (cdancy) [inline src/main/java/com/cdancy/bitbucket/rest/features/RepositoryApi.java]: Lets rename to `disableHook`

File: src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java [hunk 7]
```diff
@@ -177,6 +180,79 @@ public void testDeletePermissionByGroup() {
...
     }
 
+    @Test(dependsOnMethods = {"testGetRepository", "testCreateRepository"})
+    public void testListHook() {
+        HookPage hookPage = api().listHooks(projectKey, repoKey, 0, 100);
+        assertThat(hookPage).isNotNull();
+        assertThat(hookPage.errors()).isEmpty();
+        assertThat(hookPage.size()).isGreaterThan(0);
+        for (Hook hook : hookPage.values()) {
+            if (hook.details().configFormKey() == null) {
```
Comment (cdancy) [inline src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java]: What is the reasoning for this check?

File: src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java [hunk 8]
```diff
@@ -177,6 +180,79 @@ public void testDeletePermissionByGroup() {
...
+                hookKey = hook.details().key();
+                break;
+            }
+        }
```
Comment (cdancy) [inline src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java]: Can we add an assertion to test hookKey is not null?

File: src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java [hunk 9]
```diff
@@ -177,6 +180,79 @@ public void testDeletePermissionByGroup() {
...
+        assertThat(hook.details().key().equals(hookKey)).isTrue();
+        assertThat(hook.enabled()).isFalse();
+    }
+
+    @Test(dependsOnMethods = {"testGetRepository", "testCreateRepository"})
+    public void testDeleteHookOnError() {
+        Hook hook = api().deleteHook(projectKey, repoKey, randomStringLettersOnly() + ":" + randomStringLettersOnly());
+        assertThat(hook).isNotNull();
+        assertThat(hook.errors()).isNotEmpty();
+        assertThat(hook.enabled()).isNull();
```
Comment (cdancy) [inline src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java]: Should now be `isFalse()`

File: src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java [hunk 10]
```diff
@@ -177,6 +180,79 @@ public void testDeletePermissionByGroup() {
...
+        assertThat(hook.details().key().equals(hookKey)).isTrue();
+        assertThat(hook.enabled()).isTrue();
+    }
+
+    @Test(dependsOnMethods = {"testGetRepository", "testCreateRepository"})
+    public void testCreateHookOnError() {
+        Hook hook = api().createHook(projectKey, repoKey, randomStringLettersOnly() + ":" + randomStringLettersOnly());
+        assertThat(hook).isNotNull();
+        assertThat(hook.errors()).isNotEmpty();
+        assertThat(hook.enabled()).isNull();
```
Comment (cdancy) [inline src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java]: Should now be `isFalse()`

File: src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java [hunk 11]
```diff
@@ -177,6 +180,79 @@ public void testDeletePermissionByGroup() {
...
+        assertThat(hook.details().key().equals(hookKey)).isTrue();
+        assertThat(hook.enabled()).isFalse();
+    }
+
+    @Test(dependsOnMethods = {"testGetRepository", "testCreateRepository"})
+    public void testGetHookOnError() {
+        Hook hook = api().getHook(projectKey, repoKey, randomStringLettersOnly() + ":" + randomStringLettersOnly());
+        assertThat(hook).isNotNull();
+        assertThat(hook.errors()).isNotEmpty();
+        assertThat(hook.enabled()).isNull();
```
Comment (cdancy) [inline src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java]: Should now be `isFalse()`

File: src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java [hunk 12]
```diff
@@ -177,6 +180,79 @@ public void testDeletePermissionByGroup() {
...
+            }
+        }
+    }
+
+    @Test()
+    public void testListHookOnError() {
+        HookPage hookPage = api().listHooks(projectKey, randomString(), 0, 100);
+        assertThat(hookPage).isNotNull();
+        assertThat(hookPage.errors()).isNotEmpty();
+        assertThat(hookPage.values()).isEmpty();
```
Comment (cdancy) [inline src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java]: Should now be `isFalse()`

**Comment by cdancy** [comment]
@j0nathan33 few more comments to address. Clean these up and we can merge. Already ran tests and all worked as expected.

File: src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java (hunk 7 above)
Comment (j0nathan33) [inline src/test/java/com/cdancy/bitbucket/rest/features/RepositoryApiLiveTest.java]: If configFormKey is not null, test need to configure it before enableHook. For configurate hook,  I need to implements each rest call for all hooks.

**Comment by cdancy** [comment]
Thanks for the fixes and PR @j0nathan33 !
